lib/mayaUsd/ufe/MayaUsdUIInfoHandler.cpp

```diff
@@ -88,5 +99,68 @@ UsdUfe::UsdUIInfoHandler::SupportedTypesMap MayaUsdUIInfoHandler::getSupportedIc
     return supportedTypes;
 }
 
+Ufe::UIInfoHandler::Icon
+MayaUsdUIInfoHandler::treeViewIcon(const Ufe::SceneItem::Ptr& mayaItem) const
+{
+    Ufe::UIInfoHandler::Icon icon = Parent::treeViewIcon(mayaItem);
+
+    if (icon.baseIcon == "out_USD_Shader.png") {
+        // Naming convention for third party shader outliner icons:
+        //
+        //  We take the info:id of the shader and make it safe by replacing : with _.
+        //  Then we search the Maya icon paths for a PNG file with that name. If found we will use
+        //  it. Please note that files with _150 and _200 can also be provided for high DPI
+        //  displays.
+        //
+        //   For example an info:id of:
+        //       MyRenderer:nifty_surface
+        //   Will have this code search the full Maya icon path for a file named:
+        //       MyRenderer_nifty_surface.png
+        //   And will use it if found. At resolution 200%, the file:
+        //       MyRenderer_nifty_surface_200.png
+        //   Will alternatively be used if found.
+        //
+        const auto nodeDefHandler
+            = Ufe::RunTimeMgr::instance().nodeDefHandler(mayaItem->runTimeId());
+        if (!nodeDefHandler) {
+            return icon;
+        }
+        const auto nodeDef = nodeDefHandler->definition(mayaItem);
+        if (!nodeDef || nodeDef->type().empty()) {
+            return icon;
+        }
+
+        std::string iconName = nodeDef->type();
+        std::replace(iconName.begin(), iconName.end(), ':', '_');
+        iconName += ".png";
+
+        // Since this will be hitting the filesystem hard, mostly to find nothing, let's cache
+        // search results. Note that "XBMLANGPATH" is Maya specific, which is why the code is here
+        // and not in the base class.
```

Comment:
I understand the desire for the cache. My only comment here is that this means a user is not able to change the "info:id" of the prim and see the new icon in the Outliner. They would have to quit/restart Maya. Is that a concern?

Reply:
This cache is keyed on the file name that was searched for. It will react to a change of `info:id` because that will change the file name to lookup. Only weakness is that it is not reset when the icon path changes so there might be issues if the outliner opens before the plugin gets loaded. Do you think I should reset the cache whenever a plugin gets loaded/unloaded?

Reply:
Ah right I got a little mixed up with the cache. The icon will change if the user changes `info:id` so I think that is good enough. I wouldn't worry about the icon path changes.